cpu/stm32f3/periph/i2c.c

```diff
@@ -0,0 +1,490 @@
+static void _start(I2C_TypeDef *dev, uint8_t address, uint8_t length, uint8_t rw_flag);
+static inline void _read(I2C_TypeDef *dev, char *data, int length);
+static inline void _write(I2C_TypeDef *dev, char *data, int length);
+static inline void _stop(I2C_TypeDef *dev);
+
+/**
+ * @brief Array holding one pre-initialized mutex for each I2C device
+ */
+static mutex_t locks[] =  {
+#if I2C_0_EN
+    [I2C_0] = MUTEX_INIT,
+#endif
+#if I2C_1_EN
+    [I2C_1] = MUTEX_INIT,
+#endif
+#if I2C_2_EN
+    [I2C_2] = MUTEX_INIT
+#endif
+#if I2C_3_EN
+    [I2C_3] = MUTEX_INIT
+#endif
+};
+
+int i2c_init_master(i2c_t dev, i2c_speed_t speed)
+{
+    I2C_TypeDef *i2c;
+    GPIO_TypeDef *port_scl;
+    GPIO_TypeDef *port_sda;
+    int pin_scl = 0, pin_sda = 0;
+    int presc, scll, sclh, sdadel, scldel;
+
+    /* read speed configuration (for i2cclk = 8mhz) */
+    switch (speed) {
+        case I2C_SPEED_NORMAL:
+            presc = 1;
+            scll = 0xC7;
+            sclh = 0xC3;
+            sdadel = 0x2;
+            scldel = 0x4;
+            break;
+
+        case I2C_SPEED_FAST:
+            presc = 0;
+            scll = 0x9;
+            sclh = 0x3;
+            sdadel = 0x1;
+            scldel = 0x3;
+            break;
+
+        case I2C_SPEED_FAST_PLUS:
+            presc = 0;
+            scll = 0x6;
+            sclh = 0x3;
+            sdadel = 0x1;
+            scldel = 0x1;
+            break;
+
+        default:
+            return -2;
+    }
+
+    /* read static device configuration */
+    switch (dev) {
+#if I2C_0_EN
+        case I2C_0:
+            i2c = I2C_0_DEV;
+            port_scl = I2C_0_SCL_PORT;
+            pin_scl = I2C_0_SCL_PIN;
+            port_sda = I2C_0_SDA_PORT;
+            pin_sda = I2C_0_SDA_PIN;
+            I2C_0_CLKEN();
+            I2C_0_SCL_CLKEN();
+            I2C_0_SDA_CLKEN();
+            NVIC_SetPriority(I2C_0_ERR_IRQ, I2C_IRQ_PRIO);
+            NVIC_EnableIRQ(I2C_0_ERR_IRQ);
+            break;
+#endif
+
+        default:
+            return -1;
+    }
+
+    /* configure pins */
+    _pin_config(port_scl, port_sda, pin_scl, pin_sda);
+
+    /* configure device */
+    _i2c_init(i2c, presc, scll, sclh, sdadel, scldel);
+
+    return 0;
+}
+
+static void _i2c_init(I2C_TypeDef *i2c, int presc, int scll, int sclh, int sdadel, int scldel)
+{
+    /* disable device */
+    i2c->CR1 &= ~(I2C_CR1_PE);
+
+    /* configure analog noise filter */
+    i2c->CR1 |= I2C_CR1_ANFOFF;
+
+    /* configure digital noise filter */
+    i2c->CR1 |= I2C_CR1_DNF;
+
+    /* set timing registers */
+    i2c->TIMINGR &= ~(I2C_TIMINGR_PRESC);
+    i2c->TIMINGR |= (presc << 28);
+    i2c->TIMINGR &= ~(I2C_TIMINGR_SDADEL);
+    i2c->TIMINGR |= (sdadel << 16);
+    i2c->TIMINGR &= ~(I2C_TIMINGR_SCLDEL);
+    i2c->TIMINGR |= (scldel << 20);
+    i2c->TIMINGR &= ~(I2C_TIMINGR_SCLH);
+    i2c->TIMINGR |= (sclh << 8);
+    i2c->TIMINGR &= ~(I2C_TIMINGR_SCLL);
+    i2c->TIMINGR |= (scll);
```

Comment:
Just as an alternative way to go, what do you think: Wouldn't it be easer to create a bitmask in the upper `switch (speed) case` and set the `TIMINGR` with one code line?